htx - reorganization fetchMarkets & fetchTickers

ttodua opened the pull request:
this update reorganizes, fixes issues and simplifies huobi methods for fetching markets and tickers.
1) future and swap (linear) markets were being fetched separately, while it's possible to fetch both of them in one call per: https://www.htx.com/en-us/opend/newApiPages/?id=8cb802c2-77b5-11ed-9966-0242ac110003 (with `business_type:all` param)
2) there were inline hardcoded heavy parsing (itarations through all symbols inside fetchTickers over and over again on each ticker for all fetched tickers... very very much bad approach), also parsing each ticker was done manually inside fetchTickers, which needed to be moved into  `parseTicker`.
3) futures market-ids map is now defined once in loadMarkets, and then instantly assigned to related market-id inside parseTicker, without need of wasting hundreds of iterations for each ticker on each fetch of "fetchTickers" over and over again.
4) fetchMarketsTypes option dict is not much relevant at all, we could totally remove that option dict too, because in almost all exchanges we load all markets with loadMarkets and don't ask users which markets to load (i.e. binance, gate...), so that config is odd imo.
5) added also market-parsing logic based on returned market (i.e. if returned market is contract, instead of checking if request was made for contarct markets)

in short, when fetching futures markets : https://api.hbdm.com/linear-swap-api/v1/swap_contract_info?business_type=futures , there are only few (btc,eth) active contracts (probably huobi delisted all other second-tier coins from futures a long time ago)

so, with these updates, as you can see down below, fetchTickers correctly maps eth-usdt futures symbol, while inexistent "tomo-usdt" is left as-is (because we didn't have any info from fetchMarkets about that "tomo-usdt" futures market )




```
npm run cli.js -- htx fetchTickers undefined  '{"type":"future","subType":"linear"}'                  

> ccxt@4.2.84 cli.js
> node ./examples/js/cli.js htx fetchTickers undefined {"type":"future","subType":"linear"}

2024-03-28T10:08:20.899Z
Node.js: v20.10.0
CCXT v4.2.84
htx.fetchTickers (, [object Object])
2024-03-28T10:08:24.605Z iteration 0 passed in 874 ms

{
  'TOMO-USDT-CQ': {
    symbol: 'TOMO-USDT-CQ',
    timestamp: 1711620503779,
    datetime: '2024-03-28T10:08:23.779Z',
    high: 0.3297,
    low: 0.3297,
    open: 0.3297,
    close: 0.3297,
    last: 0.3297,
    change: 0,
    percentage: 0,
    average: 0.3297,
    baseVolume: 0,
    quoteVolume: 0,
    info: {
      id: '1711620503',
      ts: '1711620503779',
      ask: null,
      bid: null,
      trade_partition: 'USDT',
      business_type: 'futures',
      contract_code: 'TOMO-USDT-CQ',
      open: '0.3297',
      close: '0.3297',
      low: '0.3297',
      high: '0.3297',
      amount: '0',
      count: '0',
      vol: '0',
      trade_turnover: '0'
    }
  },
  'ETH-USDT-240405': {
    symbol: 'ETH-USDT-240405',
    timestamp: 1711620503779,
    datetime: '2024-03-28T10:08:23.779Z',
    high: 3618.96,
    low: 3480.12,
    bid: 3602.88,
    bidVolume: 36,
    ask: 3611.46,
    askVolume: 36,
    vwap: 100,
    open: 3549.16,
    close: 3607.68,
    last: 3607.68,
    change: 58.52,
    percentage: 1.6488408524834044,
    average: 3578.42,
    baseVolume: 3311,
    quoteVolume: 331100,
    info: {
      id: '1711620503',
      ts: '1711620503779',
      ask: [ 3611.46, 36 ],
      bid: [ 3602.88, 36 ],
      trade_partition: 'USDT',
      business_type: 'futures',
      contract_code: 'ETH-USDT-NW',
      open: '3549.16',
      close: '3607.68',
      low: '3480.12',
      high: '3618.96',
      amount: '3311',
      count: '18267',
      vol: '331100',
      trade_turnover: '11767243.5184'
    }
  },
 ...
}
2024-03-28T10:08:24.605Z iteration 1 passed in 874 ms
```

carlosmiei commented on ts/src/htx.ts:
```diff
@@ -941,16 +941,14 @@ export default class htx extends Exchange {
-                        'swap': {
-                            'linear': true,
-                            'inverse': true,
+                        'linear': true,
+                        'inverse': {
+                            'future': true,
+                            'swap': true,
```
we shouldn't do this inverts the hierarchy we have for markets type-> subtype

ttodua replied:
but this is not for hierarchy, just a dict for fetching markets

ttodua replied:
ok, i'll reorg that, will need few mins

ttodua replied:
because that logic does not either fits in our `exceptions per market type` functionality, where, i.e. for binance, we have `spot, linear, inverse ...` flat hierarchy too, instead of `spot, swap, future ...`

ttodua commented:
added also market-parsing logic based on returned market (i.e. `if returned market is contract`, instead of checking `if request was made for contarct markets`)

ttodua commented:
PR passes now, I think that fetchMarketsTypes option dict is not relevant to anyone, we could totally remove that option dict too, because in almost all exchanges we load all markets with `loadMarkets` and don't ask users which markets to load, so that config is odd imo.

ttodua commented:
I might push other PRs for htx, but at first, whenever you have a time if you can review this PR @carlosmiei , so i'll do other things only after this will be reviewed.

ttodua commented:
@carlosmiei this PR passes now

ttodua commented:
@carlosmiei any chance to review this?

carlosmiei commented on ts/src/htx.ts:
```diff
@@ -2214,39 +2252,17 @@ export default class htx extends Exchange {
+        return this.filterByArrayTickers (tickers, 'symbol', symbols);
+    }
+
+    async defineFutureMarketIdSymbols () {
```
everything about this function is confusing why do we need have something like that?

carlosmiei replied:
looks like we save 1 call in fetchMarkets but then we end up making an extra call anyway here, so it does not seem beneficial at all besides all the added complexity

ttodua commented on ts/src/htx.ts:
```diff
@@ -1614,24 +1611,21 @@ export default class htx extends Exchange {
-        const types = this.safeValue (options, 'types', {});
+        let types = undefined;
+        [ types, params ] = this.handleOptionAndParams (params, 'fetchMarkets', 'types', {});
         let allMarkets = [];
```
now it can be managed from `params` too

ttodua commented on ts/src/htx.ts:
```diff
@@ -1614,24 +1611,21 @@ export default class htx extends Exchange {
+                    promises.push (this.fetchMarketsByTypeAndSubType (undefined, 'linear', params));
+                } else if (key === 'inverse') {
+                    promises.push (this.fetchMarketsByTypeAndSubType ('swap', 'inverse', params));
+                    promises.push (this.fetchMarketsByTypeAndSubType ('future', 'inverse', params));
```
simplified market types

ttodua commented on ts/src/htx.ts:
```diff
@@ -1708,75 +1692,58 @@ export default class htx extends Exchange {
         //         "ts":1640736207263
         //     }
         //
-        // swaps
```
this was repeated block (only had few different fields, which is marked inline in proposed changes for better readability)

ttodua commented on ts/src/htx.ts:
```diff
@@ -1708,75 +1692,58 @@ export default class htx extends Exchange {
         if (numMarkets < 1) {
-            throw new NetworkError (this.id + ' fetchMarkets() returned an empty response: ' + this.json (markets));
+            throw new OperationFailed (this.id + ' fetchMarkets() returned an empty response: ' + this.json (response));
         }
```
discarded NetworkError

ttodua commented on ts/src/htx.ts:
```diff
@@ -2214,39 +2243,9 @@ export default class htx extends Exchange {
-        return this.filterByArrayTickers (result, 'symbol', symbols);
+        const rawTickers = this.safeList2 (response, 'data', 'ticks', []);
+        const tickers = this.parseTickers (rawTickers, symbols, params);
+        return this.filterByArrayTickers (tickers, 'symbol', symbols);
```
all these parsings were incorrectly happening inside `fetchTickers`, and also was incorrect code.  all the logic and symbol detection was rewritten into dedicated "trySetSymbolFromFutureMarkets"

ttodua commented on ts/src/base/Exchange.ts:
```diff
@@ -3981,7 +3981,12 @@ export default class Exchange {
+        return this.getListFromObjectValues (filteredMarkets, 'symbol');
+    }
+
+    getMarketsForMarketType (marketType: Str = undefined, subType: Str = undefined, symbolWithActiveStatus: boolean = true, symbolWithUnknownStatus: boolean = true, symbolWithInactiveStatus: boolean = true) {
```
this method was nowhere used, so we could change modify them and make better.

ttodua commented:
now, with this PR, it uses 1 less call for fetchMarkets (and neither any new call is being added anymore, I have completely simplified previous update and also simplified very much)

with master, you couldnt make calls:
```
    var a3=await e.fetchTickers (undefined, {'subType':'linear'});
```
now you can call that.

also with current master:
``` 
    var a4=await e.fetchTickers (undefined, {'type':'future', 'subType':'linear'});
    var a5=await e.fetchTickers (undefined, {'type':'future', 'subType':'inverse'});
```
you can see that symbols (BTC-USDT futures) were not resolved to markets (because "tickers" endpoints on huobi have differnt specific formated market-ids).  Their parsing is now rewritten and fixed.


also for inverse markets, you might note that resulted dict does not contain unified symbols, instead they use market-ids provided from fetch-tickers as is - that happens because during fetchMarkets those symbols are not being provided by exchange, so fetchTickers passes them as raw marketids and symbols are formed like that.
```
    var a6=await e.fetchTickers (undefined, {'subType':'inverse', 'type':'swap'});
    var a7=await e.fetchTickers (undefined, {'subType':'inverse', 'type':'future'});
```

ttodua commented:
Ready for review

carlosmiei commented on ts/src/htx.ts:
```diff
@@ -1939,6 +1919,44 @@ export default class htx extends Exchange {
+        if (symbolOrMarketId in futureMarketIdsForSymbols) {
+            return futureMarketIdsForSymbols[symbolOrMarketId];
+        }
+        const futureMarkets = this.filterBy (this.markets, 'future', true);
```
This is a bit worrying, if you call fetchTickers and you need to parse hundreds of markets then you will call this filterBy hundreds of time (which is going to slow down parsing)

ttodua replied:
if i correctly understand, 
I created a temporary dictionary (`this.options['futureMarketIdsForSymbols']`)  so in this method, that filtering happens only once per instance lifetime for any market-id, so after it is "detected" once, it's saved in `futureMarketIdsForSymbols` dict, and then in the top lines of this method, on next calls, it's quickly resolved. so, that "filter" happens only initially on first "fetchTickers" call

ttodua replied:
but you might be right - i.e. as I see now, the "undetected" symbols might be skipped from that clause, where I save marketids into temp dict. so, unsaved marketIds might cause re-filtering on next calls too. I'll push a minor commit for that too

ttodua replied:
I've pushed and now i can confirm that it's only called once on first call of `fetchTickers` and then the filter lines are never reached again.
about memory consumption, only this small dict is created: 
![2024-04-08 22 01 49](https://github.com/ccxt/ccxt/assets/7117978/171b5956-f6fb-41b0-9b3c-b9645d470b82)
